tf_quant_finance/volatility/heston_approximation.py

```diff
@@ -0,0 +1,239 @@
+# Lint as: python3
+# Copyright 2020 Google LLC
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+import math
+import tf_quant_finance as tff
+
+"""Method for semi-analytical Heston option price.
+
+Heston originally published in 1993 his eponymous model. He provided a semi-
+analytical formula for pricing European option via Fourrier transform under his
+model. However, as noted by Albrecher, the characteric function used in Heston 
+paper can suffer numerical issues because of the discontinuous nature of the 
+square root function in the complex plane, and a second version of the 
+characteric function which doesn't suffer this shortcoming should be used 
+instead. Attari further refined the numerical method by reducing the number of 
+numerical integrations (only one Fourrier transform instead of two) and with
+an integrand function with better convergence (converging in 1 / u **2 
+instead of 1 / u).  Attari's numerical method is implemented here.
+ 
+## References
+
+[1] Hansjorg Albrecher, The Little Heston Trap
+https://perswww.kuleuven.be/~u0009713/HestonTrap.pdf
+
+[2] Mukarram Attari, Option Pricing Using Fourier Transforms: A Numerically 
+Efficient Simplification
+https://papers.ssrn.com/sol3/papers.cfm?abstract_id=520042
+
+[3] Steven L. Heston, A Closed-Form Solution for Options with Stochastic 
+Volatility with Applications to Bond and Currency Options
+http://faculty.baruch.cuny.edu/lwu/890/Heston93.pdf
+"""
+
+def heston_price(*,
+                 strikes=None,
+                 expiries=None,
+                 is_call_options=None,
+                 kappas=None,
+                 thetas=None,
+                 sigmas=None,
+                 rhos=None,
+                 variances=None,
```

Comment:
Could you please make the arguments in a similar order as in [`option_price`](https://github.com/google/tf-quant-finance/blob/869a126ff1baec728091c5c2c52c13c4e9e3e059/tf_quant_finance/black_scholes/vanilla_prices.py#L20) for consistency? I.e., move `variance` to the top and forwards right after spots?
Also, is it possible to add discount_rates and dividend yield?

Reply:
I have reordered arguments a little bit. I think in general it makes sense to group on one side the option contract parameters together and on the other side the model parameters.

I have added spots, discount rates, continuous dividends like in black scholes option prices, duplicating a lot of code along the way... which is bad practise. 
I don't think it's a good idea to have lots of different ways to specify the same parameters, it just makes the code unnecessarily complicated and hard to maintain. More generally, the code should be object oriented. The conversion between spots, forwards, discount factors, dividends... should be done in one place only.